File: Golang/Graph/traverseBFS.go
package main

func bfs(graph map[string][]string, start string) []string {
	visited := make(map[string]bool)
	queue := []string{start}
	output := []string{}

	visited[start] = true

	for len(queue) > 0 {
		current := queue[0]
		queue = queue[1:] // pop front
		output = append(output, current)

		for _, neighbor := range graph[current] {
			if !visited[neighbor] {
				visited[neighbor] = true
				queue = append(queue, neighbor)
			}
		}
	}

	return output
}

File: TypeScript/Graph/traverseBFS.ts
const graph: Record<string, string[]> = {
  A: ['B', 'F'],
  B: ['A', 'F', 'C'],
  C: ['B', 'E', 'D'],
  D: ['C', 'E'],
  E: ['D', 'C', 'F'],
  F: ['A', 'B', 'E']
}

function bfs(graph: Record<string, string[]>, start: string): string[] {
  const visited = new Set<string>()
  const queue: string[] = [start]
  const output: string[] = []

  visited.add(start)

  while (queue.length > 0) {
    const current = queue.shift()! // remove first element
    output.push(current)

    for (const neighbor of graph[current]) {
      if (!visited.has(neighbor)) {
        visited.add(neighbor)
        queue.push(neighbor)
      }
    }
  }

  return output
}

console.log(bfs(graph, 'A'))